tests/components/upb/test_config_flow.py

```diff
@@ -0,0 +1,152 @@
+    assert result["data"] == {
+        "host": "serial:///dev/ttyS0:115200",
+        "file_path": "upb.upe",
+    }
+    await hass.async_block_till_done()
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_user_with_tcp_upb(hass):
+    """Test we can setup a serial upb."""
+    result = await valid_tcp_flow(hass)
+    assert result["type"] == "create_entry"
+    assert result["data"] == {"host": "tcp://1.2.3.4", "file_path": "upb.upe"}
+    await hass.async_block_till_done()
+
+
+async def test_form_cannot_connect(hass):
+    """Test we handle cannot connect error."""
+    from asyncio import TimeoutError
+
+    with patch(
+        "homeassistant.components.upb.config_flow.async_timeout.timeout",
+        side_effect=TimeoutError,
+    ):
+        result = await valid_tcp_flow(hass, sync_complete=False)
+
+    assert result["type"] == "form"
+    assert result["errors"] == {"base": "cannot_connect"}
+
+
+async def test_form_missing_upb_file(hass):
+    """Test we handle cannot connect error."""
+    result = await valid_tcp_flow(hass, config_ok=False)
+    assert result["type"] == "form"
+    assert result["errors"] == {"base": "invalid_upb_file"}
+
+
+async def test_form_user_with_already_configured(hass):
+    """Test we can setup a TCP upb."""
+    _ = await valid_tcp_flow(hass)
+    result2 = await valid_tcp_flow(hass)
+    assert result2["type"] == "abort"
+    assert result2["reason"] == "address_already_configured"
+    await hass.async_block_till_done()
+
+
+async def test_form_import(hass):
+    """Test we get the form with import source."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+
+    with mocked_upb(), patch(
+        "homeassistant.components.upb.async_setup", return_value=True
+    ) as mock_setup, patch(
+        "homeassistant.components.upb.async_setup_entry", return_value=True
+    ) as mock_setup_entry:
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": config_entries.SOURCE_IMPORT},
+            data={"host": "tcp://42.4.2.42", "file_path": "upb.upe"},
+        )
+
+    assert result["type"] == "create_entry"
+    assert result["title"] == "UPB"
+
+    assert result["data"] == {"host": "tcp://42.4.2.42", "file_path": "upb.upe"}
+    await hass.async_block_till_done()
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_junk_input(hass):
+    """Test we get the form with import source."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
+
+    with mocked_upb():
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": config_entries.SOURCE_IMPORT},
+            data={"foo": "goo", "goo": "foo"},
+        )
+
+    assert result["type"] == "form"
+    assert result["errors"] == {"base": "unknown"}
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)